server/quota/namespace.go

```diff
@@ -0,0 +1,219 @@
+	// Set by config.DefaultConfig.Quota.Namespace.Node.(Read|Write)RateLimit.
+	rateIncrement = 10
+)
+
+type Backend interface {
+	CurRates(ctx context.Context, namespace string) (int64, int64, error)
+}
+
+type instanceState struct {
+	setReadLimit  atomic.Int64
+	setWriteLimit atomic.Int64
+
+	Read  Limiter
+	Write Limiter
+}
+
+type namespace struct {
+	tenantQuota sync.Map
+	tenantMgr   *metadata.TenantManager
+
+	cfg *config.QuotaConfig
+
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	backend Backend
+}
+
+func (i *namespace) Allow(_ context.Context, namespace string, size int, isWrite bool) error {
+	s := i.getState(namespace)
+
+	us := toUnits(size, isWrite)
+
+	if isWrite {
+		return s.Write.Allow(us)
+	}
+
+	return s.Read.Allow(us)
+}
+
+func (i *namespace) Wait(ctx context.Context, namespace string, size int, isWrite bool) error {
+	s := i.getState(namespace)
+
+	us := toUnits(size, isWrite)
+
+	if isWrite {
+		return s.Write.Wait(ctx, us)
+	}
+
+	return s.Read.Wait(ctx, us)
+}
+
+func (i *namespace) getState(namespace string) *instanceState {
+	is, ok := i.tenantQuota.Load(namespace)
+	if !ok {
+		cfg := i.cfg.Namespace.Default
+		// Create new state if didn't exist before
+		is = &instanceState{
+			Write: Limiter{
+				isWrite: true,
+				Rate:    rate.NewLimiter(rate.Limit(cfg.WriteUnits), cfg.WriteUnits),
+			},
+			Read: Limiter{
+				Rate: rate.NewLimiter(rate.Limit(cfg.ReadUnits), cfg.ReadUnits),
+			},
+		}
+		i.tenantQuota.Store(namespace, is)
+	}
+
+	return is.(*instanceState)
+}
+
+func (i *namespace) loadCurNamespaceState() {
+	for _, ns := range i.tenantMgr.GetNamespaceNames() {
+		ctx1, cancel1 := context.WithTimeout(context.Background(), 5*time.Second)
+		is := i.getState(ns)
+		curRead, curWrite, err := i.backend.CurRates(ctx1, ns)
+		if err == nil {
+			i.updateLimits(ns, is, curRead, curWrite)
+		} else {
+			log.Debug().Err(err).Msg("Error updating remote rate metrics")
+		}
+		cancel1()
+	}
+}
+
+// calcLimit calculates new limit for the metric, based on updated cluster wide current and maximum rate.
+func calcLimit(setNodeLimit int64, maxNodeLimit int64, curNamespace int64, maxNamespace int64, hysteresis int64, increment int64) int64 {
+	if setNodeLimit == 0 {
+		setNodeLimit = maxNodeLimit
+	}
+
+	inc := maxNodeLimit * increment / 100
+	if inc == 0 {
+		inc = 1
+	}
+
+	if curNamespace > maxNamespace+hysteresis {
+		setNodeLimit -= inc
+	} else if curNamespace < maxNamespace-hysteresis {
+		setNodeLimit += inc
+	}
+
+	if setNodeLimit < 1 {
+		setNodeLimit = 1
+	} else if setNodeLimit > maxNodeLimit {
+		setNodeLimit = maxNodeLimit
+	}
+
+	return setNodeLimit
+}
+
+func (i *namespace) updateLimitsLow(set int64, cur int64, maxInstRate int64, maxNodeRate int) int64 {
+	return calcLimit(set, int64(maxNodeRate), cur, maxInstRate, rateHysteresis, rateIncrement)
+}
+
+func (i *namespace) updateLimits(ns string, is *instanceState, curRead int64, curWrite int64) {
```

Comment:
Probably have some metrics for the new limits, basically what are the new limits for a namespace.